### ISSUE TITLE
IRSA and AssumeRoleARN driven EKS cluster creation does not handle kubeconfig mapping correctly

### ISSUE THREAD

**Issue by rytswd** [crossplane/provider-aws#1058]
### What happened?

> Ref: Original discussion https://crossplane.slack.com/archives/C01718T2476/p1640241219051700

With Crossplane control plane cluster in AWS Account A, I have set up the following ProviderConfig for creating another cluster in AWS Account B using IRSA:

```yaml
---
apiVersion: aws.crossplane.io/v1beta1
kind: ProviderConfig
metadata:
  name: target-account-b
spec:
  assumeRoleARN: arn:aws:iam::REDACTED:role/Crossplane_Admin_Access
  credentials:
    source: InjectedIdentity
```

The cluster creation works with my basic composition with EKS. However, the following setup included in the same composition fails to apply to the cluster.

```yaml
    - name: kubernetes-provider-config
      base:
        apiVersion: kubernetes.crossplane.io/v1alpha1
        kind: ProviderConfig
        spec:
          credentials:
            source: Secret
            secretRef:
              key: kubeconfig
      patches:
        - fromFieldPath: metadata.name
          toFieldPath: metadata.name

        - fromFieldPath: spec.writeConnectionSecretToRef.namespace
          toFieldPath: spec.credentials.secretRef.namespace
        - fromFieldPath: spec.writeConnectionSecretToRef.name
          toFieldPath: spec.credentials.secretRef.name
          transforms:
            - type: string
              string:
                fmt: "%s-ekscluster"
      readinessChecks:
        - type: None
```

```yaml
    - name: aws-auth-configmap
      base:
        apiVersion: kubernetes.crossplane.io/v1alpha1
        kind: Object
        spec:
          forProvider:
            manifest:
              apiVersion: v1
              kind: ConfigMap
              metadata:
                name: aws-auth
                namespace: kube-system
      patches:
        - fromFieldPath: metadata.name
          toFieldPath: metadata.name
          transforms:
            - type: string
              string:
                fmt: "%s-eks-aws-auth"

        - type: CombineFromComposite
          combine:
            strategy: string
            variables:
              - fromFieldPath: metadata.annotations[iam.crossplane.io/account]
              - fromFieldPath: metadata.annotations[iam.crossplane.io/sso-admin-role]
            string:
              fmt: |
                - groups:
                  - system:bootstrappers
                  - system:nodes
                  rolearn: arn:aws:iam::%s:role/Crossplane_Admin_Access
                  username: system:node:{{EC2PrivateDNSName}}
                - groups:
                  - system:masters
                  rolearn: arn:aws:iam::%s:role/EKSAdminRole
                  username: eksadmin
          toFieldPath: spec.forProvider.manifest.data.mapRoles
          policy:
            fromFieldPath: Optional

        - fromFieldPath: metadata.name
          toFieldPath: spec.providerConfigRef.name
```

Error message:
`cannot create new Kubernetes client: cannot create Kubernetes client: Unauthorized`

I can see some error message in EKS cluster authenticator logs.

```log
time="2022-01-07T19:07:40Z" level=warning msg="access denied" arn="arn:aws:iam::REDACTED:role/Crossplane_Admin_Access" client="127.0.0.1:57752" error="failed mapping username: SessionName did not contain an instance id" method=POST path=/authenticate sts=sts.ap-northeast-1.amazonaws.com
```

### How can we reproduce it?

Create an EKS cluster with Crossplane composition, with the above additional setup of `ProviderConfig.kubernetes.crossplane.io` and `Object.kubernetes.crossplane.io` (I tried updating `aws-config` ConfigMap, but any resource should suffice for testing sake)

(Note there are some assumption around used names, such as `spec.writeConnectionSecretToRef.name` being `"%s-ekscluster"` with uid)

### What environment did it happen in?
Crossplane version: v1.5
provider-aws version: v0.22.0

**Comment by haarchri** [comment]
@rytswd thanks for updating in slack with
`A quick update - finally got the authentication successful with a new cluster!` - we will investigate if it is now working or something regarding an update in provider-aws to set RoleSessionName if using assumeRoleARN

**Comment by rytswd** [comment]
**TL; DR - This was a false alarm 🙇**

Apologies for the confusion above, I have managed to apply the `aws-auth` update in cross-account scenario using `provider-aws` v0.22.0. Here are some take aways.

### First update goes successfully

Because Crossplane periodically calls AWS APIs to query the actual cluster status, etc., it was difficult to figure out what happened at the "very first creation of ConfigMap" by AWS provider. I could confirm this by providing some random ConfigMap, and could see that the ConfigMap was correctly created. With that random ConfigMap, the periodic reconciliation didn't cause any issues, as it had literally no meaning at all. However, when managing `aws-auth`, it became a lot more complicated.

### Breaking `aws-auth`

The original YAML above in the issue description for `aws-auth` was essentially the root cause of all the problems. When the `aws-auth` ConfigMap under `kube-system` was updated by Crossplane once, it seemed to have gone in cleanly. However, because of the ConfigMap being essentially broken, it somehow made any further cluster manipulation became unusable, causing the second time reconciliation to fail  (i.e. `cannot create new Kubernetes client: cannot create Kubernetes client: Unauthorized`).

It was to the point that I could not even `kubectl get services` using the same IAM Role that Crossplane assumes to create the cluster. I'm not 100% sure what exactly happened, but breaking `aws-auth` was making the situation a lot more difficult to understand.

### Side note: Authentication logs

Ref: https://github.com/crossplane/provider-aws/issues/1014

When debugging, I wanted to enable Authentication logs on EKS - which caused the EKS cluster to be never in sync. This was due to the following spec not matching the API response.

```yaml
            logging:
              clusterLogging:
                - enabled: true
                  types:
                    - authenticator
```

The API response was something like below:

```yaml
            logging:
              clusterLogging:
                - enabled: true
                  types:
                    - authenticator
                - enabled: false
                  types:
                    - api
                    - audit
                    - controllerManager
                    - scheduler
```

It meant that, with the current provider, the disabled logs need to be specified in order for the cluster to be considered in sync.


## Solution

The actual spec I used in the end is something like the following. This tries to achieve the following:

- Crossplane creates the cluster in cross-account setup (e.g. `source control plane` to `target clutser`)
- Crossplane enables SSO Admin users access to Kubernetes API via `aws-auth` (this is done via custom annotation on the XR / Claim)
- Crossplane will then manage `aws-auth`, making sure any modification to it will be reverted by the periodic reconciliation


```yaml
    - name: aws-auth-configmap
      base:
        apiVersion: kubernetes.crossplane.io/v1alpha1
        kind: Object
        spec:
          forProvider:
            manifest:
              apiVersion: v1
              kind: ConfigMap
              metadata:
                name: aws-auth
                namespace: kube-system
                # The below was for testing.
                # name: aws-auth-dummy
                # namespace: default
      patches:
        - fromFieldPath: metadata.name
          toFieldPath: metadata.name
          transforms:
            - type: string
              string:
                fmt: "%s-eks-aws-auth"

        - type: CombineFromComposite
          combine:
            strategy: string
            variables:
              - fromFieldPath: metadata.annotations[iam.crossplane.io/account]
              - fromFieldPath: metadata.name
              - fromFieldPath: metadata.annotations[iam.crossplane.io/account]
              - fromFieldPath: metadata.annotations[iam.crossplane.io/sso-admin-role-name]
              - fromFieldPath: metadata.annotations[iam.crossplane.io/sso-admin-role-name]
            string:
              fmt: |
                - rolearn: arn:aws:iam::%s:role/%s-role-for-node-group
                  username: system:node:{{EC2PrivateDNSName}}
                  groups:
                  - system:bootstrappers
                  - system:nodes
                - rolearn: arn:aws:iam::%s:role/%s
                  username: %s:{{SessionName}}
                  groups:
                  - system:masters
          toFieldPath: spec.forProvider.manifest.data.mapRoles
          policy:
            fromFieldPath: Optional

        - fromFieldPath: metadata.name
          toFieldPath: spec.providerConfigRef.name
```

NOTE: With the above setup, `CombineFromComposite` is rather lengthy, and I tried to use the AWS keyword of `{{AccountID}}` to simplify the setup - but it didn't work for the RoleARN. 


---

If there are any questions, please feel free to ping me. I will also look to share the actual Composition setup once I finish finalising some more details and clean up some business specific usages.
